flocker/node/agents/ebs.py

```diff
@@ -0,0 +1,198 @@
+
+    def _is_cluster_volume(self, cluster_id, ebs_volume):
+        """
+        Helper function to check if given volume belongs to
+        given cluster.
+        """
+        actual_cluster_id = ebs_volume.tags.get(CLUSTER_ID_LABEL)
+        if actual_cluster_id is not None:
+            actual_cluster_id = UUID(actual_cluster_id)
+            if actual_cluster_id == cluster_id:
+                return True
+        return False
+
+    def create_volume(self, dataset_id, size):
+        """
+        Create a volume on EBS. Store Flocker-specific
+        {metadata version, cluster id, dataset id} for the volume
+        as volume tag data.
+
+        :param String dataset_id: Dataset_id for the volume.
+        :param int size: Requested volume size in Bytes.
+
+        :return BlockDeviceVolume volume: Created volume in
+            BlockDeviceVolume format.
+        """
+        requested_volume = self.connection.create_volume(
+            size=int(Byte(size).to_GB().value), zone=self.zone)
+
+        # Stamp created volume with Flocker-specific tags.
+        metadata = {
+            METADATA_VERSION_LABEL: '1',
+            CLUSTER_ID_LABEL: unicode(self.cluster_id),
+            DATASET_ID_LABEL: unicode(dataset_id),
+        }
+        self.connection.create_tags([requested_volume.id],
+                                    metadata)
+
+        # Wait for created volume to reach 'available' state.
+        self._wait_for_volume(requested_volume)
+
+        # Return created volume in BlockDeviceVolume format.
+        return self._blockdevicevolume_from_ebs_volume(requested_volume)
+
+    def list_volumes(self):
+        """
+        Return all volumes in {available, in-use} state that belong to
+        this Flocker cluster.
+        """
+        volumes = []
+        for ebs_volume in self.connection.get_all_volumes():
+            if ((self._is_cluster_volume(self.cluster_id, ebs_volume)) and
+               (ebs_volume.status in [u'available', u'in-use'])):
+                volumes.append(
+                    self._blockdevicevolume_from_ebs_volume(ebs_volume)
+                )
+        return volumes
+
+    def resize_volume(self, blockdevice_id, size):
+        pass
+
+    def attach_volume(self, blockdevice_id, host):
+        pass
+
+    def detach_volume(self, blockdevice_id):
+        pass
+
+    def destroy_volume(self, blockdevice_id):
+        """
+        Destroy EBS volume identified by blockdevice_id.
+
+        :param String blockdevice_id: EBS UUID for volume to be destroyed.
+
+        :raises UnknownVolume: If there does not exist a Flocker cluster
+            volume identified by input blockdevice_id.
+        :raises Exception: If we failed to destroy Flocker cluster volume
+            corresponding to input blockdevice_id.
+        """
+        for volume in self.list_volumes():
+            if volume.blockdevice_id == blockdevice_id:
+                ret_val = self.connection.delete_volume(blockdevice_id)
```

Comment:
Ok. So if I understand @exarkun 's suggestion, we should add a new IBlockDeviceAPI test to demonstrate that `delete_volume` doesn't  delete attached volumes.

Add extra logic here to prevent any attempt to delete attached volume.

And finally add Eliot message / action here to report any other failures that come back from the EBS/ / boto `delete_volume` API. (maybe the Eliot logging isn't necessary here...exceptions raised here might be logged by the Eliot logging in the calling deployer....but if we do raise an Exception here, it ought to be a BotoException and it ought to include all the potentially useful information about the failed API request.)